### PR TITLE
[MISC] Reduce memory usage due to diff constraint solver.

### LINKED ISSUE #1914: [Bug]: Memory usage increased 14+%
### Bug Description

(this is more of an FYI, feel free to close it if it's within acceptable limits)

I believe the differentiable constraint solver feature (commit 44e08794bf80c36e5cfac8544633f59392efc499) increased the memory footprint of genesis by about 14%. For simple environments, this isn't a big problem. However, for one of my slightly more [complex robots](https://github.com/jgillick/spider-bot/tree/main/simulate/genesis) it reduces the number of parallel simulations by about 1,000.

For reference, though, I'm still able to run far more parallel simulations using Genesis than Isaac Lab. So either way, your team is doing great work!

### Steps to Reproduce

Run the [Genesis locomotion example](https://github.com/Genesis-Embodied-AI/Genesis/tree/main/examples/locomotion) and check the average GPU memory usage.

**On my system**
Before this commit: 3573MiB
After this commit: 4082MiB

### Expected Behavior

The memory usage shouldn't go up quite so much. Then again, there are always tradeoffs for better features, like making the rigid body simulation to be differentiable. So maybe this is well within the expected increases as the system matures.

### Screenshots/Videos

_No response_

### Relevant log output

```shell

```

### Environment

- OS: Ubuntu 24
- GPU/CPU: RTX 5070 Ti
- GPU-driver version: 580.82.07
- CUDA / CUDA-toolkit version: 13.0 

### Release version or Commit ID

Latest `main` branch, commit 345aa3c047ae85529bc215813b8915ff4eaf4b66

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by YilingQiao**
## Description

Trying to resolve #1914 

For me, GPU memory usage in
```
python examples/locomotion/go2_train.py
```

reduce from 3.0G -> 2.6G

## Related Issue

Resolves Genesis-Embodied-AI/Genesis#<your-issue-number>

## Checklist:

- [x] I read the **CONTRIBUTING** document.
- [x] I followed the `Submitting Code Changes` section of **CONTRIBUTING** document.
- [x] I tagged the title correctly (including BUG FIX/FEATURE/MISC/BREAKING)
- [x] I updated the [documentation](https://github.com/Genesis-Embodied-AI/genesis-doc) accordingly or no change is needed.
- [x] I tested my changes and added instructions on how to test it for reviewers.

- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: genesis/engine/solvers/rigid/collider_decomp.py [hunk 1]
```diff
@@ -86,7 +86,9 @@ def _init_static_config(self) -> None:
                 geom.is_convex or geom.type == gs.GEOM_TYPE.TERRAIN for geom in self._solver.geoms
             ),
             has_terrain=any(geom.type == gs.GEOM_TYPE.TERRAIN for geom in self._solver.geoms),
-            n_contacts_per_pair=20 if self._solver._static_rigid_sim_config.requires_grad else 5,
+            # FIXME: The line below will cause error for tests/test_grad.py::test_diff_solver[cpu], why?
```
Comment (YilingQiao) [inline genesis/engine/solvers/rigid/collider_decomp.py]: when you have time, can you take a look @SonSang

Comment (YilingQiao) [reply genesis/engine/solvers/rigid/collider_decomp.py]: Since we are not using requires_grad for rigid solver, maybe we can merge this PR first to reduce the memory first for non-differentiable use

Comment (SonSang) [reply genesis/engine/solvers/rigid/collider_decomp.py]: The differentiable contact detection finds more than 5 contacts, so error should be thrown. And yes, you can go ahead. I will take care of this later again.

Comment (YilingQiao) [reply genesis/engine/solvers/rigid/collider_decomp.py]: In the test, the diff solver grad error is large, not sure why the number of contacts will impact the solver's gradient. Thanks for taking over this. This issue is not a problem until we enable all differentiable rigid solver.

Comment (SonSang) [reply genesis/engine/solvers/rigid/collider_decomp.py]: I see, yeah please go ahead. I might need to debug it more later.

Comment (duburcqa) [reply genesis/engine/solvers/rigid/collider_decomp.py]: I'm in favour of not breaking stuffs that work just to slightly reduce memory usage.

File: genesis/engine/solvers/rigid/rigid_solver_decomp.py [hunk 2]
```diff
@@ -87,6 +87,8 @@ def __init__(self, scene: "Scene", sim: "Simulator", options: RigidOptions) -> N
         self._max_collision_pairs = options.max_collision_pairs
         self._integrator = options.integrator
         self._box_box_detection = options.box_box_detection
+        # self._requires_grad = self._sim.options.requires_grad
+        self._requires_grad = True
```
Comment (duburcqa) [inline genesis/engine/solvers/rigid/rigid_solver_decomp.py]: What is this?

File: genesis/utils/array_class.py [hunk 3]
```diff
@@ -20,6 +20,10 @@
 DATA_ORIENTED = partial(dataclasses.dataclass, frozen=True) if gs.use_ndarray else ti.data_oriented
 
 
+def _s(shape, is_on):
+    return shape if is_on else (1,)
+
+
```
Comment (duburcqa) [inline genesis/utils/array_class.py]: I'm not in favour of using cryptic helpers. It just makes everything harder to read.

File: genesis/utils/array_class.py [hunk 4]
```diff
@@ -20,6 +20,10 @@
 DATA_ORIENTED = partial(dataclasses.dataclass, frozen=True) if gs.use_ndarray else ti.data_oriented
 
 
+def _s(shape, is_on):
+    return shape if is_on else (1,)
```
Comment (duburcqa) [inline genesis/utils/array_class.py]: Besides, if you want to avoid memory allocation for unused variables, you should rather use shape `()` instead of `(1,)`.

File: genesis/utils/array_class.py [hunk 5]
```diff
@@ -356,19 +358,21 @@ def get_diff_contact_input(solver, max_contacts_per_pair):
...
-        local_pos1_c=V_VEC(3, dtype=gs.ti_float, shape=(_B, max_contacts_per_pair)),
-        local_pos2_a=V_VEC(3, dtype=gs.ti_float, shape=(_B, max_contacts_per_pair)),
-        local_pos2_b=V_VEC(3, dtype=gs.ti_float, shape=(_B, max_contacts_per_pair)),
-        local_pos2_c=V_VEC(3, dtype=gs.ti_float, shape=(_B, max_contacts_per_pair)),
-        w_local_pos1=V_VEC(3, dtype=gs.ti_float, shape=(_B, max_contacts_per_pair)),
-        w_local_pos2=V_VEC(3, dtype=gs.ti_float, shape=(_B, max_contacts_per_pair)),
-        ref_id=V(dtype=gs.ti_int, shape=(_B, max_contacts_per_pair)),
-        valid=V(dtype=gs.ti_int, shape=(_B, max_contacts_per_pair)),
-        ref_penetration=V(dtype=gs.ti_float, shape=(_B, max_contacts_per_pair), needs_grad=True),
+        geom_a=V(dtype=gs.ti_int, shape=_s((_B, max_contacts_per_pair), solver._requires_grad)),
```
Comment (duburcqa) [inline genesis/utils/array_class.py]: Define `requires_grad = solver._requires_grad` to reduce line width. It would be more readable.

Comment (YilingQiao) [reply genesis/utils/array_class.py]: done

File: genesis/utils/array_class.py (hunk 4 above)
Comment (YilingQiao) [reply genesis/utils/array_class.py]: done

File: genesis/utils/array_class.py (hunk 3 above)
Comment (YilingQiao) [reply genesis/utils/array_class.py]: done, changed to `_shape`

File: genesis/engine/solvers/rigid/rigid_solver_decomp.py (hunk 2 above)
Comment (YilingQiao) [reply genesis/engine/solvers/rigid/rigid_solver_decomp.py]: it's a leftover during debug :) deleted it

File: genesis/utils/array_class.py [hunk 6]
```diff
@@ -20,6 +20,10 @@
 DATA_ORIENTED = partial(dataclasses.dataclass, frozen=True) if gs.use_ndarray else ti.data_oriented
 
 
+def _shape(shape, is_on):
```
Comment (duburcqa) [inline genesis/utils/array_class.py]: I would be better to have a more explicit name if you want to keep this pattern. Like `optional_shape` or `cond_shape` or `maybe_shape`.
